# Project 3 
## NYC Yellow Taxi Trip Analysis

In Project 1 we simply walked through the NYC Yellow Taxi Trip Data, and in this project, we will intergrate all the methods that we learned at class to give a detailed analysis on the dataset.

## 0. Setup

In [14]:
import pandas as pd
import numpy as np
import plotly.express as px
from IPython.display import HTML

pd.set_option('display.max_columns', None)
pd.set_option('display.width', 140)

data= 'yellow_taxi.csv' 

df = pd.read_csv(data)
print('Shape:', df.shape)
df.head()

Shape: (1000, 19)


vendorid     tpep_pickup_datetime    tpep_dropoff_datetime  passenger_count  trip_distance  ratecodeid store_and_fwd_flag  \
0         2  2023-01-01T00:32:10.000  2023-01-01T00:40:36.000              1.0           0.97         1.0                  N   
1         2  2023-01-01T00:55:08.000  2023-01-01T01:01:27.000              1.0           1.10         1.0                  N   
2         2  2023-01-01T00:25:04.000  2023-01-01T00:37:49.000              1.0           2.51         1.0                  N   
3         1  2023-01-01T00:03:48.000  2023-01-01T00:13:25.000              0.0           1.90         1.0                  N   
4         2  2023-01-01T00:10:29.000  2023-01-01T00:21:19.000              1.0           1.43         1.0                  N   

   pulocationid  dolocationid  payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  improvement_surcharge  total_amount  \
0           161           141             2          9.3   1.00      0.5        0.00           0.0                    1.0         14.30   
1            43           237             1          7.9   1.00      0.5        4.00           0.0                    1.0         16.90   
2            48           238             1         14.9   1.00      0.5       15.00           0.0                    1.0         34.90   
3           138             7             1         12.1   7.25      0.5        0.00           0.0                    1.0         20.85   
4           107            79             1         11.4   1.00      0.5        3.28           0.0                    1.0         19.68   

   congestion_surcharge  airport_fee  
0                   2.5         0.00  
1                   2.5         0.00  
2                   2.5         0.00  
3                   0.0         1.25  
4                   2.5         0.00

## 1. Inspecting the Raw Data

In this section we:
- Look at basic info and data types
- Check column names
- Get a first sense of missing values and basic statistics

In [6]:
df.info()

print('\nColumn names:')
print(df.columns.tolist())

print('\nMissing values per column:')
print(df.isna().sum())

df.describe(include='all').transpose().head(20)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 19 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   vendorid               1000 non-null   int64  
 1   tpep_pickup_datetime   1000 non-null   object 
 2   tpep_dropoff_datetime  1000 non-null   object 
 3   passenger_count        1000 non-null   float64
 4   trip_distance          1000 non-null   float64
 5   ratecodeid             1000 non-null   float64
 6   store_and_fwd_flag     1000 non-null   object 
 7   pulocationid           1000 non-null   int64  
 8   dolocationid           1000 non-null   int64  
 9   payment_type           1000 non-null   int64  
 10  fare_amount            1000 non-null   float64
 11  extra                  1000 non-null   float64
 12  mta_tax                1000 non-null   float64
 13  tip_amount             1000 non-null   float64
 14  tolls_amount           1000 non-null   float64
 15  impro

count unique                      top freq      mean        std   min     25%    50%    75%     max
vendorid               1000.0    NaN                      NaN  NaN     1.703   0.457165   1.0     1.0    2.0    2.0     2.0
tpep_pickup_datetime     1000    862  2023-01-01T00:27:20.000    3       NaN        NaN   NaN     NaN    NaN    NaN     NaN
tpep_dropoff_datetime    1000    881  2023-01-01T00:40:25.000    4       NaN        NaN   NaN     NaN    NaN    NaN     NaN
passenger_count        1000.0    NaN                      NaN  NaN     1.517   0.894713   0.0     1.0    1.0    2.0     6.0
trip_distance          1000.0    NaN                      NaN  NaN   2.96194   3.233793   0.0  1.1475   1.93    3.4   30.15
ratecodeid             1000.0    NaN                      NaN  NaN     1.126   3.110273   1.0     1.0    1.0    1.0    99.0
store_and_fwd_flag       1000      2                        N  996       NaN        NaN   NaN     NaN    NaN    NaN     NaN
pulocationid           1000.0    NaN                      NaN  NaN   164.838   68.55853   4.0   114.0  161.0  236.0   264.0
dolocationid           1000.0    NaN                      NaN  NaN   162.806  73.505138   4.0   107.0  161.0  236.0   265.0
payment_type           1000.0    NaN                      NaN  NaN     1.256    0.51646   1.0     1.0    1.0    1.0     4.0
fare_amount            1000.0    NaN                      NaN  NaN    16.946  13.403051 -25.4     8.6   12.8   20.5   140.2
extra                  1000.0    NaN                      NaN  NaN    1.7315   1.246279  -1.0     1.0    1.0    3.5    9.75
mta_tax                1000.0    NaN                      NaN  NaN     0.493   0.080358  -0.5     0.5    0.5    0.5     0.5
tip_amount             1000.0    NaN                      NaN  NaN   3.24868   3.432172   0.0     0.0   2.85    4.4    40.0
tolls_amount           1000.0    NaN                      NaN  NaN   0.15381   1.113337   0.0     0.0    0.0    0.0    18.3
improvement_surcharge  1000.0    NaN                      NaN  NaN    0.9845   0.161975  -1.0     1.0    1.0    1.0     1.0
total_amount           1000.0    NaN                      NaN  NaN  25.19374  16.155035 -30.4   15.48  20.52  29.76  189.98
congestion_surcharge   1000.0    NaN                      NaN  NaN    2.2975   0.726823  -2.5     2.5    2.5    2.5     2.5
airport_fee            1000.0    NaN                      NaN  NaN   0.05375   0.253698   0.0     0.0    0.0    0.0    1.25

## 2. Cleaning & Feature Engineering

Typical NYC yellow taxi data includes columns like:
- `tpep_pickup_datetime`, `tpep_dropoff_datetime`
- `passenger_count`
- `trip_distance`
- `fare_amount`, `tip_amount`, `total_amount`
- Pickup / dropoff location IDs

We will:
1. Standardize column names (lower-case, strip spaces)
2. Convert pickup and dropoff times to `datetime`
3. Convert key numeric columns (distance, fares, passengers)
4. Create new features:
   - trip duration (minutes)
   - pickup hour of day
   - pickup weekday
   - pickup date

In [7]:
# Standardize column names
df.columns = [c.strip().lower() for c in df.columns]
print('Standardized columns:')
print(df.columns.tolist())

# Helper to find a column by possible names
def find_col(possible_names):
    for name in df.columns:
        for cand in possible_names:
            if name == cand.lower():
                return name
    return None

pickup_col = find_col(['tpep_pickup_datetime', 'pickup_datetime', 'pickup_time'])
dropoff_col = find_col(['tpep_dropoff_datetime', 'dropoff_datetime', 'dropoff_time'])
passenger_col = find_col(['passenger_count'])
dist_col = find_col(['trip_distance'])
fare_col = find_col(['fare_amount'])
tip_col = find_col(['tip_amount'])
total_col = find_col(['total_amount'])

print('\nDetected key columns:')
for label, col in [
    ('pickup', pickup_col),
    ('dropoff', dropoff_col),
    ('passenger_count', passenger_col),
    ('trip_distance', dist_col),
    ('fare_amount', fare_col),
    ('tip_amount', tip_col),
    ('total_amount', total_col),
]:
    print(f'{label:15s}: {col}')

Standardized columns:
['vendorid', 'tpep_pickup_datetime', 'tpep_dropoff_datetime', 'passenger_count', 'trip_distance', 'ratecodeid', 'store_and_fwd_flag', 'pulocationid', 'dolocationid', 'payment_type', 'fare_amount', 'extra', 'mta_tax', 'tip_amount', 'tolls_amount', 'improvement_surcharge', 'total_amount', 'congestion_surcharge', 'airport_fee']

Detected key columns:
pickup         : tpep_pickup_datetime
dropoff        : tpep_dropoff_datetime
passenger_count: passenger_count
trip_distance  : trip_distance
fare_amount    : fare_amount
tip_amount     : tip_amount
total_amount   : total_amount


In [8]:
# Convert datetime columns
if pickup_col is not None:
    df[pickup_col] = pd.to_datetime(df[pickup_col], errors='coerce')
if dropoff_col is not None:
    df[dropoff_col] = pd.to_datetime(df[dropoff_col], errors='coerce')

# Convert numeric columns
numeric_candidates = [passenger_col, dist_col, fare_col, tip_col, total_col]
for col in numeric_candidates:
    if col is not None:
        df[col] = pd.to_numeric(df[col], errors='coerce')

# Create engineered time features
if pickup_col is not None:
    df['pickup_datetime'] = df[pickup_col]
    df['pickup_date'] = df[pickup_col].dt.date
    df['pickup_hour'] = df[pickup_col].dt.hour
    df['pickup_weekday'] = df[pickup_col].dt.dayofweek  # 0=Mon, 6=Sun

# Trip duration in minutes
if (pickup_col is not None) and (dropoff_col is not None):
    df['trip_duration_min'] = (df[dropoff_col] - df[pickup_col]).dt.total_seconds() / 60

df.head()

vendorid tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  trip_distance  ratecodeid store_and_fwd_flag  pulocationid  \
0         2  2023-01-01 00:32:10   2023-01-01 00:40:36              1.0           0.97         1.0                  N           161   
1         2  2023-01-01 00:55:08   2023-01-01 01:01:27              1.0           1.10         1.0                  N            43   
2         2  2023-01-01 00:25:04   2023-01-01 00:37:49              1.0           2.51         1.0                  N            48   
3         1  2023-01-01 00:03:48   2023-01-01 00:13:25              0.0           1.90         1.0                  N           138   
4         2  2023-01-01 00:10:29   2023-01-01 00:21:19              1.0           1.43         1.0                  N           107   

   dolocationid  payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  improvement_surcharge  total_amount  \
0           141             2          9.3   1.00      0.5        0.00           0.0                    1.0         14.30   
1           237             1          7.9   1.00      0.5        4.00           0.0                    1.0         16.90   
2           238             1         14.9   1.00      0.5       15.00           0.0                    1.0         34.90   
3             7             1         12.1   7.25      0.5        0.00           0.0                    1.0         20.85   
4            79             1         11.4   1.00      0.5        3.28           0.0                    1.0         19.68   

   congestion_surcharge  airport_fee     pickup_datetime pickup_date  pickup_hour  pickup_weekday  trip_duration_min  
0                   2.5         0.00 2023-01-01 00:32:10  2023-01-01            0               6           8.433333  
1                   2.5         0.00 2023-01-01 00:55:08  2023-01-01            0               6           6.316667  
2                   2.5         0.00 2023-01-01 00:25:04  2023-01-01            0               6          12.750000  
3                   0.0         1.25 2023-01-01 00:03:48  2023-01-01            0               6           9.616667  
4                   2.5         0.00 2023-01-01 00:10:29  2023-01-01            0               6          10.833333

### 2.1 Basic Sanity Filters

Here we can apply simple filters to remove clearly invalid trips, such as:
- Non-positive or extremely large distances
- Non-positive or extremely long trip durations
- Negative fares

In [9]:
df_clean = df.copy()

# Filter on distance (e.g., 0 < distance < 100 miles)
if dist_col is not None:
    df_clean = df_clean[(df_clean[dist_col] > 0) & (df_clean[dist_col] < 100)]

# Filter on duration (e.g., 0 < duration < 240 minutes)
if 'trip_duration_min' in df_clean.columns:
    df_clean = df_clean[(df_clean['trip_duration_min'] > 0) & (df_clean['trip_duration_min'] < 240)]

# Filter on fare (e.g., fare > 0)
if fare_col is not None:
    df_clean = df_clean[df_clean[fare_col] > 0]

print('Original rows:', len(df))
print('Rows after basic filters:', len(df_clean))

Original rows: 1000
Rows after basic filters: 983


## 3. Exploratory Data Analysis (EDA)

### 3.1 Distributions of Key Numeric Variables

We look at basic distributions:
- Trip distance
- Trip duration (if available)
- Fare amount
- Tip and total amounts

In [22]:
def px_hist(col, nbins=50, title=None):
    if col is None or col not in df_clean.columns:
        print(f'Skipping {col}: not found.')
        return
    if title is None:
        title = f'Distribution of {col}'
    fig = px.histogram(df_clean, x=col, nbins=nbins, title=title)
    fig.show()
    HTML(fig.to_html(include_plotlyjs="cdn", full_html=False))

px_hist(dist_col, nbins=50, title=f'Distribution of {dist_col}')
if 'trip_duration_min' in df_clean.columns:
    px_hist('trip_duration_min', nbins=50, title='Distribution of Trip Duration (min)')
px_hist(fare_col, nbins=50, title='Distribution of Fare Amount')
px_hist(tip_col, nbins=50, title='Distribution of Tip Amount')
px_hist(total_col, nbins=50, title='Distribution of Total Amount')

All these histograms are positively skewed which totally matches the real-life common sense that short taxi trips are more common (expecially during midnight when trains are out of service).

### 3.2 Temporal Patterns

We examine how demand (number of trips) changes by:
- Hour of day
- Day of week

This is useful for understanding rush hours and weekday vs weekend patterns.

In [ ]:
# Trips by pickup hour
if 'pickup_hour' in df_clean.columns:
    trips_by_hour = df_clean.groupby('pickup_hour').size().reset_index(name='count')
    fig = px.bar(trips_by_hour, x='pickup_hour', y='count',
                 title='Trips by Pickup Hour')
    fig.show()
    HTML(fig.to_html(include_plotlyjs="cdn", full_html=False))

# Trips by weekday
if 'pickup_weekday' in df_clean.columns:
    trips_by_wd = df_clean.groupby('pickup_weekday').size().reset_index(name='count')
    fig = px.bar(trips_by_wd, x='pickup_weekday', y='count',
                 title='Trips by Weekday (0=Mon, 6=Sun)')
    fig.show()
    HTML(fig.to_html(include_plotlyjs="cdn", full_html=False))

As we can derive from the two histograms that the data only includes a short time window over the New Year's Eve in 2023 (from 2022.12.31 to 2023.1.1), despite it's unable to figure out what time of the day or what day of the week that we have the most passengers, we could still use this data to see:
1. When did the number of order started to surge?
2. Whether the passengers tend to tip more after midnight?

To answer the two questions we could use the following codes:

In [18]:
df_clean['pickup_minute'] = df_clean[pickup_col].dt.floor('T')

minute_counts = df_clean.groupby('pickup_minute').size().reset_index(name='count')

fig = px.line(minute_counts, x='pickup_minute', y='count',
              title='Trips per Minute (High-Resolution Demand)')
fig.show()
HTML(fig.to_html(include_plotlyjs="cdn", full_html=False))

/var/folders/bj/gpmqq0dx5ld8w1w42kl86m2h0000gn/T/ipykernel_59581/2486904844.py:1: FutureWarning:

'T' is deprecated and will be removed in a future version, please use 'min' instead.



In [19]:
df_clean['tip_rate'] = df_clean[tip_col] / df_clean[fare_col]

fig = px.histogram(df_clean, x='tip_rate', nbins=40,
                   title='Distribution of Tip Rate')
fig.show()
HTML(fig.to_html(include_plotlyjs="cdn", full_html=False))

In [20]:
df_clean['is_after_midnight'] = df_clean[pickup_col].dt.hour >= 1

fig = px.box(df_clean, x='is_after_midnight', y='tip_rate',
             labels={'is_after_midnight': 'After Midnight'},
             title='Tip Rate: Before vs After Midnight')
fig.show()
HTML(fig.to_html(include_plotlyjs="cdn", full_html=False))

## A Small Sum Up
1. The minute-level pickup distribution exhibits a clear event-driven pattern: a pre-midnight buildup, an abrupt dip at exactly midnight, and a strong post-celebration rebound. This behavior is highly characteristic of large-scale public events such as New Year’s Eve in NYC, where mobility freezes at the celebration peak and resumes sharply as crowds begin dispersing.
2. Tip rates follow a positively skewed distribution with a dominant cluster between 10% and 30%, consistent with normative tipping conventions in NYC. A small number of outliers exceeding 100% likely reflect celebratory generosity or short-fare trips where fixed-amount tips inflate the percentage.
3. A clear behavioral shift emerges across midnight: passengers before midnight tip both more generously and with greater variability, while post-midnight tipping becomes more conservative and tightly distributed. This pattern aligns with consumer psychology research showing elevated prosocial spending during emotionally salient moments (such as the NYE countdown), followed by a reversion to more utilitarian decision-making once the event concludes.

### 3.3 Relationship between Distance and Fare

We expect longer trips to cost more. Here we visualize the relationship between
trip distance and fare amount using an interactive scatter plot.

In [ ]:
if (dist_col is not None) and (fare_col is not None):
    fig = px.scatter(df_clean, x=dist_col, y=fare_col, opacity=0.3,
                     title='Fare Amount vs Trip Distance')
    fig.show()
    HTML(fig.to_html(include_plotlyjs="cdn", full_html=False))

    corr = df_clean[[dist_col, fare_col]].corr().iloc[0, 1]
    print(f'Correlation between distance and fare: {corr:.3f}')

Correlation between distance and fare: 0.955


The scatter plot shows a very strong positive correlation between distance and fare (r ≈ 0.955):
	•	Base fare + distance-based + time-based components produce a near-linear relationship.
	•	Deviations above the trend reflect congestion or prolonged idling, which is expected on NYE.
	•	Very long-distance trips, though fewer, remain consistent with the meter structure.

## 5. Conclusion

Together, the visualizations paint a coherent picture:
	1.	Demand is highly time-sensitive and event-shaped, responding immediately to crowd behavior around the NYE countdown.
	2.	Passenger decision-making shifts at midnight, showing an emotional-to-practical behavioral transition.
	3.	Generosity and tipping patterns are influenced by the social context, not just fare size or distance.
	4.	Taxi fare economics remain structurally consistent, even during high-congestion, high-activity periods.

Overall, even in a small time window, New Year’s Eve taxi data clearly illustrates how urban mobility, economic behavior, and crowd psychology interact—showing predictable fare structure, emotionally influenced tipping, and sharply event-driven demand patterns.